From 8f8ae4b80f42b98edc261d8b6f4aa808cc88682d Mon Sep 17 00:00:00 2001
From: Zubair Lutfullah <zubair.lutfullah@gmail.com>
Date: Sun, 30 Jun 2013 19:08:03 +0100
Subject: [PATCH 21/21] Add IIO map interface for bone-helper Most of the work
 comes from Pantelis Antoniou's patch in the original
 patchset. This adds the IIO map interface which is
 used by bone-helper. I didnt want to disturb existing
 users of the interface. The values match the ones in
 /sys/bus/iio/*/in_voltage*_scale for DT vsense-scale
 values of 100.

---
 drivers/iio/adc/ti_am335x_adc.c |   28 +++++++++++++++++++++++++++-
 1 file changed, 27 insertions(+), 1 deletion(-)

diff --git a/drivers/iio/adc/ti_am335x_adc.c b/drivers/iio/adc/ti_am335x_adc.c
index c030078..f16cac9 100644
--- a/drivers/iio/adc/ti_am335x_adc.c
+++ b/drivers/iio/adc/ti_am335x_adc.c
@@ -36,6 +36,8 @@ struct tiadc_device {
 	int channels;
 	u8 channel_line[8];
 	u8 channel_step[8];
+	char *buf;
+	struct iio_map *map;
 };
 
 static unsigned int tiadc_readl(struct tiadc_device *adc, unsigned int reg)
@@ -106,7 +108,7 @@ static int tiadc_channel_init(struct iio_dev *indio_dev, int channels)
 	struct tiadc_device *adc_dev = iio_priv(indio_dev);
 	struct iio_chan_spec *chan_array;
 	struct iio_chan_spec *chan;
-	int i;
+	int i,ret;
 
 	indio_dev->num_channels = channels;
 	chan_array = kcalloc(channels,
@@ -126,10 +128,34 @@ static int tiadc_channel_init(struct iio_dev *indio_dev, int channels)
 		chan->scan_type.sign = 'u';
 		chan->scan_type.realbits = 12;
 		chan->scan_type.storagebits = 32;
+		chan->scan_type.shift = 0;
 	}
 
 	indio_dev->channels = chan_array;
 
+	adc_dev->map = kcalloc(indio_dev->num_channels + 1,sizeof(struct iio_map), GFP_KERNEL);
+
+	if (adc_dev->map == NULL) {
+		kfree(chan_array);
+		return -ENOMEM;
+	}
+
+	for (i = 0; i < indio_dev->num_channels; i++) {
+		adc_dev->map[i].adc_channel_label = chan_array[i].datasheet_name;
+		adc_dev->map[i].consumer_dev_name = "any";
+		adc_dev->map[i].consumer_channel = chan_array[i].datasheet_name;
+	}
+	adc_dev->map[i].adc_channel_label = NULL;
+	adc_dev->map[i].consumer_dev_name = NULL;
+	adc_dev->map[i].consumer_channel = NULL;
+
+	ret = iio_map_array_register(indio_dev, adc_dev->map);
+	if (ret != 0) {
+		kfree(adc_dev->map);
+		kfree(chan_array);
+		return -ENOMEM;
+	}
+
 	return 0;
 }
 
-- 
1.7.9.5

